### PR TITLE
JVM_IR: mark inline lambdas and callable references with special origins

### PR DESCRIPTION
**Pull request by pyos**
Keeping them as LOCAL_FUNCTION_FOR_LAMBDA is confusing and not very useful. The new origins make locating them much easier, and also normalize handling of lambdas and anonymous functions.

 #KT-48319 Fixed
 #KT-47279 Fixed?
 #KT-48478 Fixed

### CODE REVIEW

**Comment by pyos** [comment]
This is also necessary for fixing KT-48478: turns out the assumption that adapted function references are tail-call is incorrect - they're not if the adapter is not inline due to possibility of Unit return type coercion.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InlineCallableReferenceToLambda.kt [hunk 1]
```diff
@@ -43,158 +45,126 @@ internal val inlineCallableReferenceToLambdaPhase = makeIrFilePhase(
...
-        return if (expression.field?.owner == null) {
-            // Use getter if field is absent ...
-            expandInlineFunctionReferenceToLambda(expression, expression.getter!!.owner)
-        } else {
-            // ... else use field itself
-            expandInlineFieldReferenceToLambda(expression, expression.field!!.owner)
+    private fun IrExpression.transform(scope: IrDeclaration?) = when {
+        this is IrBlock && origin.isInlinable -> apply {
+            // Already a lambda or similar, just mark it with an origin.
+            val reference = statements[statements.size - 1] as IrFunctionReference
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InlineCallableReferenceToLambda.kt:73]: Minor: statements.lastIndex

**Comment by max-kammerer** [comment]
Please mute `IrCodegenBoxWasmTestGenerated$CallableReference$AdaptedReferences.testSuspendUnitConversion`:
```
  kotlin.NotImplementedError: An operation is not implemented.
    at org.jetbrains.kotlin.backend.wasm.WasmSymbols.getSuspendCoroutineUninterceptedOrReturn(WasmSymbols.kt:63)
    at org.jetbrains.kotlin.backend.wasm.WasmSymbols.getSuspendCoroutineUninterceptedOrReturn(WasmSymbols.kt:26)
```

**Comment by max-kammerer** [comment]
@ilmirus Please take a look

**Review by ilmirus — APPROVED** [review]

**Comment by pyos** [comment]
@max-kammerer ping?

**Comment by max-kammerer** [comment]
@pyos Sorry for delay with review and thanks for PR!

**Pull request merged**